tests/templates/test_subroutines/test_basis_rotation.py

```diff
@@ -0,0 +1,512 @@
+                        [0.62651582 + 0.0j, -0.00828925 - 0.60570321j, -0.36704948 + 0.32528067j],
+                        [-0.58608928 + 0.0j, 0.03902657 + 0.04633548j, -0.57220635 + 0.57044649j],
+                    ]
+                ),
+                [
+                    ([0], 2.2707802713289267),
+                    ([0, 1], 2.9355948424220206),
+                    ([1], -1.4869222527726533),
+                    ([1, 2], 1.2601662579297865),
+                    ([0], 2.3559705032936717),
+                    ([0, 1], 1.1748572730890159),
+                ],
+                [([0], 2.2500537657656356), ([1], -0.7251404204443089), ([2], 2.3577346350335198)],
+            ),
+        ],
+    )
+    def test_basis_rotation_operations(self, num_wires, unitary_matrix, givens, diags):
+        """Test the correctness of the BasisRotation template including the gate count
+        and their order, the wires the operation acts on and the correct use of parameters
+        in the circuit."""
+
+        gate_ops, gate_angles, gate_wires = [], [], []
+
+        for (indices, angle) in diags + givens[::-1]:
+            g_op = qml.PhaseShift if len(indices) == 1 else qml.SingleExcitation
+            gate_ops.append(g_op)
+            gate_angles.append(qml.numpy.array(angle))
+            gate_wires.append(list(indices))
+
+        op = qml.BasisRotation(wires=range(num_wires), unitary_matrix=unitary_matrix)
+        queue = op.expand().operations
+
+        assert len(queue) == len(gate_ops)  # number of gates
+
+        for idx, op in enumerate(queue):
+            assert isinstance(op, gate_ops[idx])  # gate operation
+            assert np.allclose(op.parameters[0], gate_angles[idx])  # gate parameter
+            assert list(op.wires) == gate_wires[idx]  # gate wires
+
+    def test_custom_wire_labels(self, tol):
+        """Test that BasisRotation template can deal with non-numeric, nonconsecutive wire labels."""
+
+        weights = qml.math.array(
+            [
+                [-0.618452, -0.68369054 - 0.38740723j],
+                [-0.78582258, 0.53807284 + 0.30489424j],
+            ]
+        )
+
+        dev = qml.device("default.qubit", wires=2)
+        dev2 = qml.device("default.qubit", wires=["z", "a"])
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.PauliX(wires=[0])
+            qml.BasisRotation(
+                wires=range(2),
+                unitary_matrix=weights,
+            )
+            return qml.expval(qml.Identity(0))
+
+        @qml.qnode(dev2)
+        def circuit2():
+            qml.PauliX(wires=["z"])
+            qml.BasisRotation(
+                wires=["z", "a"],
+                unitary_matrix=weights,
+            )
+            return qml.expval(qml.Identity("z"))
+
+        circuit()
+        circuit2()
+
+        assert np.allclose(dev.state, dev2.state, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize(
+        ("unitary_matrix", "eigen_values", "exp_state"),
```

Comment:
How the reference `exp_state` is computed?

Reply:
It was computed numerically via `scipy.linalg.expm`.